src/dbnode/persist/fs/commitlog/writer.go

```diff
@@ -295,27 +304,205 @@ func (w *writer) Close() error {
 	return nil
 }
 
-func (w *writer) write(data []byte) error {
+func (w *writer) write(data []byte, callbackFn callbackFn) error {
 	dataLen := len(data)
 	sizeLen := binary.PutUvarint(w.sizeBuffer, uint64(dataLen))
-	totalLen := sizeLen + dataLen
 
-	// Avoid writing across the checksum boundary if we can avoid it
-	if w.buffer.Buffered() > 0 && totalLen > w.buffer.Available() {
-		if err := w.buffer.Flush(); err != nil {
-			return err
+	// Write size and then data
+	if err := w.buffer.write(w.sizeBuffer[:sizeLen], nil); err != nil {
+		return err
+	}
+	return w.buffer.write(data, callbackFn)
+}
+
+type asyncFlushBuffer struct {
+	writers            []asyncFlushBufferWriter
+	singleBufferLength int
+	idx                int
+	flushCalls         chan asyncFlushBufferCall
+	chunkWriter        chunkWriter
+}
+
+type asyncFlushBufferWriter struct {
+	buffer    []byte
+	callbacks []callbackFn
+	result    *asyncFlushBufferCallResult
+}
+
+type asyncFlushBufferCallType uint
+
+const (
+	asyncFlushBufferFlushCall asyncFlushBufferCallType = iota
+	asyncFlushBufferSyncCall
+)
+
+type asyncFlushBufferCall struct {
+	bytes     []byte
+	callbacks []callbackFn
+	syncWg    *sync.WaitGroup
+	callType  asyncFlushBufferCallType
+	result    *asyncFlushBufferCallResult
+}
+
+type asyncFlushBufferCallResult struct {
+	wg  *sync.WaitGroup
+	n   int
+	err error
+}
+
+func (r *asyncFlushBufferCallResult) reset() {
+	r.n = 0
+	r.err = nil
+}
+
+type asyncFlushBufferOptions struct {
+	totalBufferLength  int
+	singleBufferLength int
+}
+
+func newAsyncFlushBuffer(
+	chunkWriter chunkWriter,
+	opts asyncFlushBufferOptions,
+) *asyncFlushBuffer {
+	var (
+		writers []asyncFlushBufferWriter
+		total   int
+	)
+	for total+opts.singleBufferLength <= opts.totalBufferLength {
```

Comment:
nit; can just use regular for here `for total := 0 ; total < opts.totalBufferLength; total += opts.singleBufferLength {`?